### ISSUE TITLE
Tooltips cannot be dismissed on iOS

### ISSUE THREAD

**Issue by upngo** [twbs/bootstrap#16028]
Tooltips can be shown on mobile by clicking, but you cant get rid of them.

They swallow up click events too so you can't add click handlers to the body to hide them.

Same as the problem noted in #3417 #6232 - both full of exasperated users..

If this isn't going to be fixed, at least a note could be put in the docs saying don't use tooltips if you want it to work on mobile without custom code disabling them for mobile devices...

**Comment by cvrebert** [comment]
You're referring only to iOS and not Android, correct?

We could give [`body { cursor: pointer; }`](https://developer.mozilla.org/en-US/docs/Web/Events/click#Safari_Mobile) a try.

**Comment by upngo** [comment]
Yes. That would make it more intuitive for iOS users but the problem is you
can't get them to disappear after the tap on iOS

**Comment by cvrebert** [comment]
Uh, no, I mean that should literally theoretically fix the dismissibility problem. See linked article.

**Comment by upngo** [comment]
Woops! True, that's convenient. Good find

**Comment by fat** [comment]
@cvrebert is this a css issue?

**Comment by cvrebert** [comment]
@fat It's an iOS Safari quirk/bug: https://developer.mozilla.org/en-US/docs/Web/Events/click#Safari_Mobile

**Comment by fat** [comment]
@cvrebert i just meant more, that if we wanted to pursue a fix, woudl we just add that css snippet to the tooltip component… or did you expect the js to fix this somehow? (label wise)

**Comment by cvrebert** [comment]
@fat I guess it would need to be in the JS. Something like:
```js
if ('ontouchstart' in document.documentElement) { // or whatever "is this a touch device?" test we want to use
  $('body').css('cursor', 'pointer');
}
```

**Comment by patrickhlauke** [comment]
> `or whatever "is this a touch device?" test we want to use`

there is no bullet-proof way for this part though...and it would likely cause quirky effects on hybrid devices like touch-enabled laptops, surface, etc

**Comment by cvrebert** [comment]
I guess we'd ought to toss in some iOS Safari detection/UA-sniffing for good measure.

**Comment by kkirsche** [comment]
@cvrebert Why not just use the `Touch` object safari defines on mobile?

```javascript
// Safari Mobile
if(window.Touch) { /* JavaScript for your touch interface */ }
```

**Comment by cvrebert** [comment]
@kkirsche I'm pretty sure that's not specific enough. There are browsers besides Safari that support Touch Events.

**Comment by kkirsche** [comment]
Ah, ok. Thank you for the information. I wasn't aware of that :smiley:

**Comment by deksden** [comment]
If JavaScript will fix this bug, it would be cool to have auto-hiding tooltips (after 2sec for ex).

**Comment by patrickhlauke** [comment]
auto-hiding would potentially cause issues for keyboard (and particularly assistive technology) users, depending on who it's implemented...

**Comment by deksden** [comment]
@patrickhlauke : problems are related with keyboard be hiding when DOM changes? 

Maybe if we can preserve input focus and did not reset viewport - all will be ok?

**Comment by patrickhlauke** [comment]
I'd need to see an example implementation to give a definitive yes/no, but in principle: what happens if a user moves focus to something which then triggers a tooltip. Will they have enough time to read it before it disappears. And with assistive technologies, will AT still announce the content if it disappears suddenly. Even for sighted mouse users, will they have enough time to see the tooltip before it disappears again. etc

**Comment by beej75** [comment]
I see this is still a problem in v4 as well in iOS.

I don't know if it's worth mentioning, but whatever they're doing in this jquery plugin over here - it works in iOS. It ain't smooth (unless that's just the crappy iPhone 4S I was testing on), but at least it works:
http://swisnl.github.io/jQuery-contextMenu/demo/trigger-left-click.html

**Comment by stereotipo** [comment]
solution http://jsfiddle.net/stereotipo/suhq88b0/17/

test on iOS (ipad and iphone), android and windows

**Comment by twbs-lmvtfy** [comment]
Hi @stereotipo!

You appear to have posted a live example (http://fiddle.jshell.net/stereotipo/suhq88b0/17/show/light/), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 96, column 362 thru column 368: Stray end tag ```` span ````.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by stereotipo** [comment]
solution http://jsfiddle.net/stereotipo/suhq88b0/18/
test on iOS (ipad and iphone), android and windows

**Comment by twbs-lmvtfy** [comment]
Hi @stereotipo!

You appear to have posted a live example (http://fiddle.jshell.net/stereotipo/suhq88b0/18/show/light/), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 94, column 3: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.5 ; saw what appears to be usage of Bootstrap 3.1.1

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by fat** [comment]
oh yeah… i think the best solution imho for ios is using an invisible screen… surprised i didn't already do that in v4

**Comment by cvrebert** [comment]
Turns out it's not possible to make the `<body>` clickable in iOS Safari using either `cursor:pointer` or `onclick="void(0)"`. You can make its descendants clickable, but not the `<body>` itself. This means e.g. tapping on the empty space between top-level `<p>`s or `<div>`s wouldn't dismiss tooltips, which would make for kinda weird, suboptimal UX.
Filed https://bugs.webkit.org/show_bug.cgi?id=151933 regarding this weirdness.
Agreed that a backdrop looks like the only viable option.
@fat Is switching to an invisible screen approach within scope or out of scope for v3?

**Comment by gregblass** [comment]
My current workaround (#page wraps all my content):

```
function closePopovers() {
  $('.popover').popover('hide');
  $("#page").unbind("click", closePopovers); 
};

// Popovers
$('[data-toggle="popover"]').popover({html: true});
  
// Bootstrap popovers don't close on outside click
$('[data-toggle="popover"]').on('shown.bs.popover', function () {
  setTimeout(function() {
    $("#page").bind("click", closePopovers); 
  }, 0);
});

Now I need to figure out how to get around the double click bug after a hide.

**Comment by gregblass** [comment]
For people looking for an alternative solution that closes on an outside click by default and doesn't require two clicks after that, this worked perfectly out of the box for me with the default options:

https://github.com/sandywalker/webui-popover

**Comment by TeTiRoss** [comment]
@stereotipo in iOS when open tooltip and then close it by click somewhere outside, the only way to open closed tooltip again is to open another tooltip and then first one will be able to open

**Comment by sleepydisaster** [comment]
The issue is still reproducing on iOS devices. Is it going to be fixed in any of the nearest releases? Does anybody know a good solution until the library is updated ?

**Comment by gregblass** [comment]
@sleepydisaster Yep, check out my solution posted above - https://github.com/sandywalker/webui-popover - currently using that instead in my production app and it works great.

**Comment by henwen** [comment]
@gregblass managed to get around the multiple tooltip issue with this code (uses [isMobile](https://github.com/kaimallea/isMobile/) to run for iOS-only):

```
  if (isMobile.apple.device) {
    // isolate the fix to apply only when a tooltip is shown
    $("[data-toggle="tooltip"]").on("shown.bs.tooltip", function() {
      $('body').css('cursor', 'pointer');
    });

    // Return cursor to default value when tooltip hidden.
    // Use 'hide' event (as opposed to "hidden") to ensure .css() fires before
    // 'shown' event for another tooltip.
    $("[data-toggle="tooltip"]").on("hide.bs.tooltip", function() {
      $('body').css('cursor', 'auto');
    });
  }
```

**Comment by mdo** [comment]
[**Bootstrap 3 is no longer being officially developed or supported.**](https://github.com/twbs/bootstrap/issues/20631)

All work has moved onto our next major release, v4. As such, this issue or pull request is being closed as a "won't fix." For additional help and support, we recommend [utilizing our community resources](https://github.com/twbs/bootstrap/blob/master/README.md#community). Thanks for your understanding, and see you on the other side of v4!

<3,
@mdo and team

**Comment by cvrebert** [comment]
Reopening since this also has the v4 label.

**Comment by alekseyg** [comment]
I recently had this issue and resolved it by using touch events and ignoring mouse events that come immediately after said touch events. I used Bootstrap's `trigger: 'manual'` option when initializing to avoid hacks. This solution uses hover on desktop, but can be expanded to work with click, focus, etc. There are a few issues with the solution: 1) on some Android devices, the touch events tend to fire unreliably, and 2) on Opera Mini, you can still touch through the tooltip, which is undesirable, but would require a 360ms delay when closing the tooltip to work around. On Android, 10ms was sufficient.

[Here is a gist with the solution](https://gist.github.com/alekseyg/d2261e42a9b5335a8ce0774b4d92f129). Let me know if this is something that could be adapted as an official solution, if there are any major flaws with it, and/or if there's really a much better way of doing things already brewing. I'd like to be able to help with resolving this issue and would definitely like to improve what I have as well.

**Comment by rrfranco92** [comment]
@alekseyg I confirm your solution is working today at version alpha 5.

Just used it in order to fix this issue and got it working.

**Comment by bashhack** [comment]
@alekseyg - just chiming in to second that confirmation of success with the solution - thank you!

**Comment by stelco** [comment]
this worked for me...
`            $('[data-toggle="popover"]').popover();
            // close on body click
            // iOS doesnt recognise 'body' click so using :not
            $(':not(#anything)').on('click', function (e) {
                $('[data-toggle="popover"]').each(function () {
                    if (!$(this).is(e.target) && $(this).has(e.target).length === 0 && $('.popover').has(e.target).length === 0) {
                        $(this).popover('hide');
                    }
                });
            });`

**Comment by stephenbe** [comment]
Actually probably the reason why the invisible screen was never implemented is probably because this doesn't make sens if you have a clickable element in the page, you will need to click twice to focus into an input...I guess it's going to be a waiting game on IOS at this point =/

**Comment by bkdotcom** [comment]
> probably the reason why the invisible screen was never implemented is probably because this doesn't make sens if you have a clickable element in the page, you will need to click twice to focus

counter-intuitive, but would 
`pointer-events: none;`
on the screen solve the problem?

**Comment by jlavoie13** [comment]
@alekseyg Is there a way to update your code so it doesn't close on scroll?

**Comment by alekseyg** [comment]
@jlavoie13 I've run into several tiny issues with this approach and have been working on improving it. I'll update the gist soon with some improvements.

**Comment by patrickhlauke** [comment]
see the contortions one has to go through for iOS here https://github.com/twbs/bootstrap/pull/22426

i'd really like to avoid having an invisible backdrop, as that just causes new problems (on touch-enabled devices with additional inputs, like touchscreen laptops/tablets with paired mouse).

i really wish iOS followed what Android and co do, which is to NOT have weird quirks relating to event bubbling/delegation...

**Comment by alekseyg** [comment]
@jlavoie13 I've updated my gist. Here's a summary of what I changed:
- The `wasTouchedRecently` function used event timestamps to calculate when a touch happened. Unfortunately, different browsers use different timestamps, so it was unreliable. I replaced it with a data attribute and a timeout for better reliability.
- I added `e.preventDefault()` on `touchend` of the element with the tooltip and on the body if a tooltip is shown to prevent tap-through actions from taking place when showing or hiding a tooltip. This also allowed me to get rid of the delay when hiding the tooltip.
- I added a `wasTapped` check on the body in order to differentiate between taps and other gestures, like when toggling the tooltip. This should prevent scrolling from hiding the tooltip.

**Comment by patrickhlauke** [comment]
not tested this yet, but as a note: adding a `touchstart` to the `body` indiscriminately will lead to bad scroll performance (since the browser can then only scroll sync rather than async, and can't thread), unless using passive event listeners (which removes the problem, but only in modern supporting browsers)

**Comment by patrickhlauke** [comment]
#22481 ports the same hack i used for dropdowns to tooltips. while hacky, it's relatively small and has no known side-effects - all it does is coax iOS into doing event delegation/bubbling like *all other touch-enabled browsers* (grumble grumble)

**Comment by PlPadraig** [comment]
@gregblass I checked out https://github.com/sandywalker/webui-popover and it worked for me as well right out of the box! Thanks! That saved me a lot of time in resolving this problem in my application.

**Comment by turigeza** [comment]
I ended up with this so far so good ... 
`
@supports (-webkit-overflow-scrolling: touch) {
    /* CSS specific to iOS devices */ 
    body{
        cursor:pointer;
    }
}`

**Comment by Joyrex** [comment]
While @turigeza solution works so tooltips are dismissible, this is only for iOS Safari, and in cases where a button invokes a modal or similar, the tooltip remains until you tap elsewhere.

I really think a solution that targets touch-enabled devices (not just iOS Safari) that auto-dismisses a tooltip after [x] seconds would be the best, and with Bootstrap 4 being mobile-first, I think should be part of the framework.

**Comment by patrickhlauke** [comment]
so let's be clear...is there still a current issue even after fixes etc? if so, please file a fresh issue with a reduced test case

**Comment by chrisbo** [comment]
I agree with Joyrex. I have just removed every single tooltip from a site because I noticed in iOS they stack and won't go away.

**Comment by Joyrex** [comment]
@patrickhlauke - well, yes - there really never was an adequate “fix” to begin with, for the reasons I pointed out.

**Comment by patrickhlauke** [comment]
and again, i'll ask you to post a reduced test case to show the issue, and post a fresh issue. the workaround implemented at the time covered the common scenario of tooltips like http://getbootstrap.com/docs/4.1/components/tooltips/#examples and they work fine on iOS. if there's further situations that currently fail...post a new issue instead of adding to a closed thread started in 2015

**Comment by BonBonSlick** [comment]
On iOS after click tooltip does not disappear BS 3.

**Comment by kiler129** [comment]
@turigeza: FYI, this no longer works :(

**Comment by Johann-S** [comment]
If you have an issue please create a new one with a live test on a CodePen thanks
